File: packages/node/src/webhooks.ts
import { createPublicKey, verify as cryptoVerify, type KeyObject } from 'node:crypto'
import { WebhookVerificationError } from './errors'
import type { WebhookEvent } from './types'

/**
 * Verification of Catalisa Webhooks Engine deliveries, exactly as the engine signs them
 * (`src/webhooks-engine/utils/crypto.ts` and `services/signing-key.service.ts`):
 *
 *   message   = x-webhook-id + "\n" + x-webhook-timestamp + "\n" + raw body
 *   signature = RSA-SHA256 (PKCS#1 v1.5) with the subscription's private key, base64
 *   header    = x-webhook-signature: "v1=<base64>"; x-webhook-key-id selects the public key
 *
 * The signature is ASYMMETRIC: you verify with the subscription's PUBLIC key
 * (`GET /webhooks-engine/api/v1/subscriptions/:id/keys`), not with a shared secret.
 * The server does not reject old deliveries — the receiver enforces the time tolerance (default 300 s).
 */

/** Accepted public keys: keyId → PEM map, `{ keyId, publicKey }` list, or a single PEM. */
export type WebhookPublicKeys =
  | string
  | Record<string, string>
  | ReadonlyArray<{ keyId: string; publicKey: string }>

export type HeadersLike =
  | Headers
  | Record<string, string | string[] | undefined>

export interface VerifyOptions {
  /** Timestamp tolerance in seconds. Default 300. `0` disables it (not recommended). */
  toleranceSeconds?: number
  /** Reference clock (tests). */
  now?: Date | number
}

export type WebhookFailure =
  | 'MISSING_HEADERS'
  | 'TIMESTAMP_INVALID'
  | 'TIMESTAMP_OUT_OF_TOLERANCE'
  | 'UNKNOWN_KEY_ID'
  | 'UNSUPPORTED_SIGNATURE_VERSION'
  | 'INVALID_SIGNATURE'

const MESSAGES: Record<WebhookFailure, string> = {
  MISSING_HEADERS: 'Missing x-webhook-id, x-webhook-timestamp, x-webhook-key-id or x-webhook-signature header',
  TIMESTAMP_INVALID: 'x-webhook-timestamp is not an ISO 8601 date',
  TIMESTAMP_OUT_OF_TOLERANCE: 'Delivery is outside the time tolerance (possible replay)',
  UNKNOWN_KEY_ID: 'No public key for this x-webhook-key-id',
  UNSUPPORTED_SIGNATURE_VERSION: 'Unsupported signature version (expected v1=)',
  INVALID_SIGNATURE: 'Invalid signature',
}

function header(headers: HeadersLike, name: string): string | undefined {
  if (typeof (headers as Headers).get === 'function') return (headers as Headers).get(name) ?? undefined
  const rec = headers as Record<string, string | string[] | undefined>
  let v = rec[name]
  if (v === undefined) {
    const key = Object.keys(rec).find((k) => k.toLowerCase() === name)
    v = key ? rec[key] : undefined
  }
  return Array.isArray(v) ? v[0] : v
}

const keyCache = new Map<string, KeyObject>()
function keyFor(keys: WebhookPublicKeys, keyId: string): KeyObject | null {
  let pem: string | undefined
  if (typeof keys === 'string') pem = keys
  else if (Array.isArray(keys)) pem = keys.find((k) => k.keyId === keyId)?.publicKey
  else pem = Object.prototype.hasOwnProperty.call(keys, keyId) ? (keys as Record<string, string>)[keyId] : undefined
  if (!pem) return null
  let k = keyCache.get(pem)
  if (!k) {
    k = createPublicKey(pem)
    keyCache.set(pem, k)
  }
  return k
}

function toBytes(raw: string | Uint8Array): Buffer {
  return typeof raw === 'string' ? Buffer.from(raw, 'utf8') : Buffer.from(raw.buffer, raw.byteOffset, raw.byteLength)
}

/** Diagnostics: `null` when valid; otherwise the failure reason. */
export function checkWebhook(rawBody: string | Uint8Array, headers: HeadersLike, keys: WebhookPublicKeys, opts: VerifyOptions = {}): WebhookFailure | null {
  const id = header(headers, 'x-webhook-id')
  const timestamp = header(headers, 'x-webhook-timestamp')
  const keyId = header(headers, 'x-webhook-key-id')
  const signature = header(headers, 'x-webhook-signature')
  if (!id || !timestamp || !keyId || !signature) return 'MISSING_HEADERS'

  const sentAt = Date.parse(timestamp)
  if (!Number.isFinite(sentAt)) return 'TIMESTAMP_INVALID'
  const tolerance = opts.toleranceSeconds ?? 300
  const now = opts.now === undefined ? Date.now() : typeof opts.now === 'number' ? opts.now : opts.now.getTime()
  if (tolerance > 0 && Math.abs(now - sentAt) > tolerance * 1000) return 'TIMESTAMP_OUT_OF_TOLERANCE'

  const key = keyFor(keys, keyId)
  if (!key) return 'UNKNOWN_KEY_ID'

  if (!signature.startsWith('v1=')) return 'UNSUPPORTED_SIGNATURE_VERSION'
  const sig = Buffer.from(signature.slice(3), 'base64')
  const message = Buffer.concat([Buffer.from(`${id}\n${timestamp}\n`, 'utf8'), toBytes(rawBody)])
  let ok = false
  try {
    ok = cryptoVerify('sha256', message, key, sig)
  } catch {
    ok = false
  }
  return ok ? null : 'INVALID_SIGNATURE'
}

/** `true` when the delivery is authentic and within tolerance. Never throws on bad signatures. */
export function verify(rawBody: string | Uint8Array, headers: HeadersLike, keys: WebhookPublicKeys, opts?: VerifyOptions): boolean {
  return checkWebhook(rawBody, headers, keys, opts) === null
}

/**
 * Verifies and returns the event. Throws `WebhookVerificationError` (with `code` = reason) when it
 * is not authentic. Pass the RAW body — re-serializing the JSON changes the bytes and breaks the signature.
 */
export function constructEvent<T = Record<string, unknown>>(
  rawBody: string | Uint8Array,
  headers: HeadersLike,
  keys: WebhookPublicKeys,
  opts?: VerifyOptions
): WebhookEvent<T> {
  const failure = checkWebhook(rawBody, headers, keys, opts)
  if (failure) throw new WebhookVerificationError(MESSAGES[failure], { status: 400, code: failure })
  const text = typeof rawBody === 'string' ? rawBody : toBytes(rawBody).toString('utf8')
  try {
    return JSON.parse(text) as WebhookEvent<T>
  } catch {
    throw new WebhookVerificationError('Webhook body is not JSON', { status: 400, code: 'INVALID_JSON' })
  }
}

File: packages/ruby/lib/catalisa/biometrics/webhooks.rb
# frozen_string_literal: true

require "base64"
require "json"
require "openssl"
require "time"

module Catalisa
  module Biometrics
    # Verification of Catalisa Webhooks Engine deliveries:
    #
    #   message   = x-webhook-id + "\n" + x-webhook-timestamp + "\n" + raw body
    #   signature = RSA-SHA256 (PKCS#1 v1.5), base64, in x-webhook-signature as "v1=<base64>"
    #   key       = the subscription's PUBLIC key, chosen by x-webhook-key-id
    #
    # The signature is asymmetric: there is no shared secret to leak. The engine
    # does not refuse old deliveries — the receiver enforces the time tolerance,
    # or a captured delivery could be replayed forever.
    class WebhookVerifier
      MESSAGES = {
        missing_headers: "missing x-webhook-id, x-webhook-timestamp, x-webhook-key-id or x-webhook-signature",
        timestamp_invalid: "x-webhook-timestamp is not an ISO 8601 date",
        timestamp_out_of_tolerance: "delivery is outside the time tolerance (possible replay)",
        unknown_key_id: "no public key for this x-webhook-key-id",
        unsupported_signature_version: "unsupported signature version (expected v1=)",
        invalid_signature: "invalid signature",
      }.freeze

      # @param public_keys [Hash{String=>String}] keyId => PEM. Keep a retired key
      #   here while deliveries signed with it may still arrive.
      # @param tolerance [Integer] seconds; 0 disables the check (not recommended)
      # @param now [Proc] reference clock, for tests
      def initialize(public_keys, tolerance: 300, now: -> { Time.now })
        @public_keys = public_keys
        @tolerance = tolerance
        @now = now
      end

      # Verifies a delivery and returns the parsed event.
      #
      # Pass the RAW body — the bytes as they arrived. Decoding and re-encoding
      # the JSON changes them, and the signature will not match.
      #
      # @param headers [Hash] however your stack hands them over: Rack's
      #   HTTP_X_WEBHOOK_ID, a framework's x-webhook-id, arrays included
      def verify(headers, raw_body)
        id = header(headers, "x-webhook-id")
        timestamp = header(headers, "x-webhook-timestamp")
        key_id = header(headers, "x-webhook-key-id")
        signature = header(headers, "x-webhook-signature")

        fail_with(:missing_headers) if [id, timestamp, key_id, signature].any? { |v| v.nil? || v.empty? }

        sent_at = begin
          Time.iso8601(timestamp)
        rescue ArgumentError
          fail_with(:timestamp_invalid)
        end
        fail_with(:timestamp_out_of_tolerance) if @tolerance.positive? && (@now.call - sent_at).abs > @tolerance
        fail_with(:unsupported_signature_version) unless signature.start_with?("v1=")

        pem = @public_keys[key_id]
        fail_with(:unknown_key_id) if pem.nil?

        raw = Base64.strict_decode64(signature[3..]) rescue fail_with(:invalid_signature)
        key = begin
          OpenSSL::PKey::RSA.new(pem)
        rescue OpenSSL::PKey::RSAError
          fail_with(:unknown_key_id)
        end

        message = "#{id}\n#{timestamp}\n#{raw_body}"
        fail_with(:invalid_signature) unless key.verify(OpenSSL::Digest.new("SHA256"), raw, message)

        begin
          JSON.parse(raw_body)
        rescue JSON::ParserError => e
          fail_with(:invalid_signature, "the signed body is not valid JSON: #{e.message}")
        end
      end

      private

      def header(headers, name)
        headers.each do |key, value|
          normalized = key.to_s.downcase.tr("_", "-")
          normalized = normalized.delete_prefix("http-")
          next unless normalized == name

          found = value.is_a?(Array) ? value.first : value
          return found.nil? || found.to_s.empty? ? nil : found.to_s
        end
        nil
      end

      def fail_with(failure, message = nil)
        raise WebhookVerificationError.new(failure, message || MESSAGES.fetch(failure))
      end
    end

    # OFFLINE verification of the signed evidence, without calling Catalisa:
    #
    #   message   = "{sessionId}|{attempt}|{bundleHash}|{signedAt}"
    #   signature = Ed25519 (RFC 8032), base64
    #   key       = GET /evidence-keys, by the signature's keyId
    #
    # This is what lets a third party — an auditor, a court — confirm years later
    # that the bundle is the one Catalisa signed.
    module Evidence
      module_function

      # The exact string that was signed.
      def message(session_id, attempt, bundle_hash, signed_at)
        "#{session_id}|#{attempt}|#{bundle_hash}|#{signed_at}"
      end

      # @return [Hash] valid:, key_id:, unknown_key:, retired_at:
      def verify(session_id, attempt, bundle_hash, signature, keys)
        key_id = signature["keyId"] || signature[:keyId]
        key = keys.find { |k| (k["keyId"] || k[:keyId]) == key_id }
        return { valid: false, key_id: key_id, unknown_key: true, retired_at: nil } if key.nil?

        retired_at = key["retiredAt"] || key[:retiredAt]
        out = { valid: false, key_id: key_id, unknown_key: false, retired_at: retired_at }
        return out unless (signature["alg"] || signature[:alg]) == "Ed25519"

        raw = Base64.strict_decode64((signature["value"] || signature[:value]).to_s) rescue (return out)
        public_key = begin
          OpenSSL::PKey.read(key["publicKeyPem"] || key[:publicKeyPem])
        rescue OpenSSL::PKey::PKeyError
          return out
        end
        # Ed25519 signs the message itself, with no separate digest step: hence the
        # nil digest. Not verify_raw — that one answers false for Ed25519 instead
        # of raising, which is a quiet way to reject every valid receipt.
        signed = message(session_id, attempt, bundle_hash, signature["signedAt"] || signature[:signedAt])
        out[:valid] = public_key.verify(nil, raw, signed)

        out
      rescue OpenSSL::PKey::PKeyError
        out
      end
    end
  end
end
